File: CS/Form1.cs
// Developer Express Code Central Example:
// How to provide the ability to resize rows
// 
// Description: I am trying to provide my end-users with the ability to change a
// specific row's height via the mouse. Is there a way to implement this feature?
// Solution: The XtraGrid doesn’t provide such a feature automatically due to the
// fact that it doesn't cache any data internally. Thus it can’t store different
// row sizes. However, you can implement this task yourself. To do this you should
// handle the GridView's MouseDown, MouseUp and CalcRowHeight events. Within the
// MouseDown and the MouseUp events you should handle row size changing. Within the
// CalcRowHeight event handler you should pass the valid row height for a specific
// row. In the attached project you will find a sample project which demonstrates
// this task
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E617

using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;

namespace Grid3
{
	/// <summary>
	/// Summary description for Form1.
	/// </summary>
	public class Form1 : System.Windows.Forms.Form
	{
        private DevExpress.XtraGrid.GridControl gridControl1;
        private DevExpress.XtraGrid.Views.Grid.GridView gridView1;
		/// <summary>
		/// Required designer variable.
		/// </summary>
		private System.ComponentModel.Container components = null;

		public Form1()
		{
			//
			// Required for Windows Form Designer support
			//
			InitializeComponent();

			//
			// TODO: Add any constructor code after InitializeComponent call
			//
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if (components != null) 
				{
					components.Dispose();
				}
			}
			base.Dispose( disposing );
		}

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
            this.gridControl1 = new DevExpress.XtraGrid.GridControl();
            this.gridView1 = new DevExpress.XtraGrid.Views.Grid.GridView();
            ((System.ComponentModel.ISupportInitialize)(this.gridControl1)).BeginInit();
            ((System.ComponentModel.ISupportInitialize)(this.gridView1)).BeginInit();
            this.SuspendLayout();
            // 
            // gridControl1
            // 
            this.gridControl1.Dock = System.Windows.Forms.DockStyle.Fill;
            // 
            // gridControl1.EmbeddedNavigator
            // 
            this.gridControl1.EmbeddedNavigator.Name = "";
            this.gridControl1.Location = new System.Drawing.Point(0, 0);
            this.gridControl1.LookAndFeel.SkinName = "Money Twins";
            this.gridControl1.LookAndFeel.Style = DevExpress.LookAndFeel.LookAndFeelStyle.Skin;
            this.gridControl1.LookAndFeel.UseDefaultLookAndFeel = false;
            this.gridControl1.LookAndFeel.UseWindowsXPTheme = false;
            this.gridControl1.MainView = this.gridView1;
            this.gridControl1.Name = "gridControl1";
            this.gridControl1.Size = new System.Drawing.Size(776, 422);
            this.gridControl1.TabIndex = 0;
            this.gridControl1.UseEmbeddedNavigator = true;
            this.gridControl1.ViewCollection.AddRange(new DevExpress.XtraGrid.Views.Base.BaseView[] {
                                                                                                        this.gridView1});
            // 
            // gridView1
            // 
            this.gridView1.GridControl = this.gridControl1;
            this.gridView1.Name = "gridView1";
            this.gridView1.OptionsCustomization.AllowRowSizing = true;
            this.gridView1.CalcRowHeight += new DevExpress.XtraGrid.Views.Grid.RowHeightEventHandler(this.gridView1_CalcRowHeight);
            this.gridView1.MouseDown += new System.Windows.Forms.MouseEventHandler(this.gridView1_MouseDown);
            this.gridView1.MouseUp += new System.Windows.Forms.MouseEventHandler(this.gridView1_MouseUp);
            // 
            // Form1
            // 
            this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
            this.ClientSize = new System.Drawing.Size(776, 422);
            this.Controls.Add(this.gridControl1);
            this.Name = "Form1";
            this.Text = "Form1";
            this.Load += new System.EventHandler(this.Form1_Load);
            ((System.ComponentModel.ISupportInitialize)(this.gridControl1)).EndInit();
            ((System.ComponentModel.ISupportInitialize)(this.gridView1)).EndInit();
            this.ResumeLayout(false);

        }
		#endregion

		/// <summary>
		/// The main entry point for the application.
		/// </summary>
		[STAThread]
		static void Main() 
		{
			Application.Run(new Form1());
		}

        Hashtable rowsSizes;
        const int defaultRowHeight = 30;
        
        private void Form1_Load(object sender, System.EventArgs e) {
            rowsSizes = new Hashtable();
            InitData();
            FillData();
            InitGridColumns();
            InitDataEvents();
        }

        //<gridControl1>
        private DataTable data;
        private void InitData() {
            data = new DataTable("ColumnsTable");
            data.BeginInit();
            AddColumn(data, "ID", System.Type.GetType("System.Int32"), true);
            AddColumn(data, "First Name", System.Type.GetType("System.String"));
            AddColumn(data, "Last Name", System.Type.GetType("System.String"));
            AddColumn(data, "Payment Type", System.Type.GetType("System.String"));
            AddColumn(data, "Customer", System.Type.GetType("System.Boolean"));
            AddColumn(data, "Payment Amount", System.Type.GetType("System.Single"));
            data.EndInit();
        }
        //</gridControl1>

        private void AddColumn(DataTable data, string name, System.Type type) { AddColumn(data, name, type, false); } 
        private void AddColumn(DataTable data, string name, System.Type type, bool ro) {
            DataColumn col;
            col = new DataColumn(name, type);
            col.Caption = name;
            col.ReadOnly = ro;
            data.Columns.Add(col);
        }

        private void FillData() {
            string[,] sNames = new string[,] { {
                                                   "Elizabeth", "Lincoln"}, {"Yang", "Wang"}, { 
                                                                                                  "Patricio", "Simpson"}, {"Francisco", "Chang"}, { 
                                                                                                                                                      "Ann", "Devon"}, {"Roland", "Mendel"}, { 
                                                                                                                                                                                                 "Paolo", "Accorti"}, {"Diego", "Roel"}}; 
            string[] sType = new string[] {"Visa", "Master", "Cash"};
            data.Clear();
            Random rnd = new Random();
            for(int i = 0; i <= sNames.GetUpperBound(0); i++) 
                data.Rows.Add(new object[] {i + 1, sNames[i,0], sNames[i,1], sType[i % 3], rnd.Next(-1, 1), rnd.Next(10000) * 0.01});
        }

        private void InitGridColumns() {
            gridControl1.DataSource = data;
            gridControl1.DefaultView.PopulateColumns();
            gridView1.Columns["Payment Amount"].DisplayFormat.FormatType = DevExpress.Utils.FormatType.Numeric;
            gridView1.Columns["Payment Amount"].DisplayFormat.FormatString = "c";
            gridView1.BestFitColumns();
        }

        private void InitDataEvents() {
            data.RowDeleting += new DataRowChangeEventHandler(data_RowDeleting); 
        }

        private void data_RowDeleting(object sender, DataRowChangeEventArgs e) {
            rowsSizes.Remove(e.Row);
        }
        
        private void gridView1_CalcRowHeight(object sender, DevExpress.XtraGrid.Views.Grid.RowHeightEventArgs e) {
            object rowHeight = rowsSizes[gridView1.GetDataRow(e.RowHandle)];
            if (rowHeight != null)  
                e.RowHeight = Convert.ToInt32(rowHeight);
            else
                e.RowHeight = defaultRowHeight;
        }

        int resizeStartPos;
        int resizingRowHandle = -1;

        private void gridView1_MouseDown(object sender, System.Windows.Forms.MouseEventArgs e) {
            DevExpress.XtraGrid.Views.Grid.ViewInfo.GridHitInfo  hi = gridView1.CalcHitInfo(e.X, e.Y);
            if (hi.HitTest == DevExpress.XtraGrid.Views.Grid.ViewInfo.GridHitTest.RowEdge) {
                resizeStartPos = e.Y;
                resizingRowHandle = hi.RowHandle;
            }
        }

        private void gridView1_MouseUp(object sender, System.Windows.Forms.MouseEventArgs e) {
            if (resizingRowHandle >= 0) {
                resizeStartPos = e.Y - resizeStartPos;
                object rowSize = rowsSizes[gridView1.GetDataRow(resizingRowHandle)];
                int newRowSize;
                if (rowSize != null) 
                    newRowSize = Convert.ToInt32(rowSize);
                else
                    newRowSize = defaultRowHeight;
                newRowSize += resizeStartPos;
                if (newRowSize < defaultRowHeight)
                    newRowSize = defaultRowHeight;
                rowsSizes[gridView1.GetDataRow(resizingRowHandle)] = newRowSize;
                resizingRowHandle = -1;
            }
        }

    }
}

File: VB/Form1.vb
' Developer Express Code Central Example:
' How to provide the ability to resize rows
' 
' Description: I am trying to provide my end-users with the ability to change a
' specific row's height via the mouse. Is there a way to implement this feature?
' Solution: The XtraGrid doesn’t provide such a feature automatically due to the
' fact that it doesn't cache any data internally. Thus it can’t store different
' row sizes. However, you can implement this task yourself. To do this you should
' handle the GridView's MouseDown, MouseUp and CalcRowHeight events. Within the
' MouseDown and the MouseUp events you should handle row size changing. Within the
' CalcRowHeight event handler you should pass the valid row height for a specific
' row. In the attached project you will find a sample project which demonstrates
' this task
' 
' You can find sample updates and versions for different programming languages here:
' http://www.devexpress.com/example=E617


Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Data

Namespace Grid3
	''' <summary>
	''' Summary description for Form1.
	''' </summary>
	Public Class Form1
		Inherits System.Windows.Forms.Form
		Private gridControl1 As DevExpress.XtraGrid.GridControl
		Private WithEvents gridView1 As DevExpress.XtraGrid.Views.Grid.GridView
		''' <summary>
		''' Required designer variable.
		''' </summary>
		Private components As System.ComponentModel.Container = Nothing

		Public Sub New()
			'
			' Required for Windows Form Designer support
			'
			InitializeComponent()

			'
			' TODO: Add any constructor code after InitializeComponent call
			'
		End Sub

		''' <summary>
		''' Clean up any resources being used.
		''' </summary>
		Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
			If disposing Then
				If components IsNot Nothing Then
					components.Dispose()
				End If
			End If
			MyBase.Dispose(disposing)
		End Sub

		#Region "Windows Form Designer generated code"
		''' <summary>
		''' Required method for Designer support - do not modify
		''' the contents of this method with the code editor.
		''' </summary>
		Private Sub InitializeComponent()
			Me.gridControl1 = New DevExpress.XtraGrid.GridControl()
			Me.gridView1 = New DevExpress.XtraGrid.Views.Grid.GridView()
			CType(Me.gridControl1, System.ComponentModel.ISupportInitialize).BeginInit()
			CType(Me.gridView1, System.ComponentModel.ISupportInitialize).BeginInit()
			Me.SuspendLayout()
			' 
			' gridControl1
			' 
			Me.gridControl1.Dock = System.Windows.Forms.DockStyle.Fill
			' 
			' gridControl1.EmbeddedNavigator
			' 
			Me.gridControl1.EmbeddedNavigator.Name = ""
			Me.gridControl1.Location = New System.Drawing.Point(0, 0)
			Me.gridControl1.LookAndFeel.SkinName = "Money Twins"
			Me.gridControl1.LookAndFeel.Style = DevExpress.LookAndFeel.LookAndFeelStyle.Skin
			Me.gridControl1.LookAndFeel.UseDefaultLookAndFeel = False
			Me.gridControl1.LookAndFeel.UseWindowsXPTheme = False
			Me.gridControl1.MainView = Me.gridView1
			Me.gridControl1.Name = "gridControl1"
			Me.gridControl1.Size = New System.Drawing.Size(776, 422)
			Me.gridControl1.TabIndex = 0
			Me.gridControl1.UseEmbeddedNavigator = True
			Me.gridControl1.ViewCollection.AddRange(New DevExpress.XtraGrid.Views.Base.BaseView() { Me.gridView1})
			' 
			' gridView1
			' 
			Me.gridView1.GridControl = Me.gridControl1
			Me.gridView1.Name = "gridView1"
			Me.gridView1.OptionsCustomization.AllowRowSizing = True
'			Me.gridView1.CalcRowHeight += New DevExpress.XtraGrid.Views.Grid.RowHeightEventHandler(Me.gridView1_CalcRowHeight);
'			Me.gridView1.MouseDown += New System.Windows.Forms.MouseEventHandler(Me.gridView1_MouseDown);
'			Me.gridView1.MouseUp += New System.Windows.Forms.MouseEventHandler(Me.gridView1_MouseUp);
			' 
			' Form1
			' 
			Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
			Me.ClientSize = New System.Drawing.Size(776, 422)
			Me.Controls.Add(Me.gridControl1)
			Me.Name = "Form1"
			Me.Text = "Form1"
'			Me.Load += New System.EventHandler(Me.Form1_Load);
			CType(Me.gridControl1, System.ComponentModel.ISupportInitialize).EndInit()
			CType(Me.gridView1, System.ComponentModel.ISupportInitialize).EndInit()
			Me.ResumeLayout(False)

		End Sub
		#End Region

		''' <summary>
		''' The main entry point for the application.
		''' </summary>
		<STAThread> _
		Shared Sub Main()
			Application.Run(New Form1())
		End Sub

		Private rowsSizes As Hashtable
		Private Const defaultRowHeight As Integer = 30

		Private Sub Form1_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Load
			rowsSizes = New Hashtable()
			InitData()
			FillData()
			InitGridColumns()
			InitDataEvents()
		End Sub

		'<gridControl1>
		Private data As DataTable
		Private Sub InitData()
			data = New DataTable("ColumnsTable")
			data.BeginInit()
			AddColumn(data, "ID", System.Type.GetType("System.Int32"), True)
			AddColumn(data, "First Name", System.Type.GetType("System.String"))
			AddColumn(data, "Last Name", System.Type.GetType("System.String"))
			AddColumn(data, "Payment Type", System.Type.GetType("System.String"))
			AddColumn(data, "Customer", System.Type.GetType("System.Boolean"))
			AddColumn(data, "Payment Amount", System.Type.GetType("System.Single"))
			data.EndInit()
		End Sub
		'</gridControl1>

		Private Sub AddColumn(ByVal data As DataTable, ByVal name As String, ByVal type As System.Type)
			AddColumn(data, name, type, False)
		End Sub
		Private Sub AddColumn(ByVal data As DataTable, ByVal name As String, ByVal type As System.Type, ByVal ro As Boolean)
			Dim col As DataColumn
			col = New DataColumn(name, type)
			col.Caption = name
			col.ReadOnly = ro
			data.Columns.Add(col)
		End Sub

		Private Sub FillData()
			Dim sNames(,) As String = { { "Elizabeth", "Lincoln"}, {"Yang", "Wang"}, { "Patricio", "Simpson"}, {"Francisco", "Chang"}, { "Ann", "Devon"}, {"Roland", "Mendel"}, { "Paolo", "Accorti"}, {"Diego", "Roel"}}
			Dim sType() As String = {"Visa", "Master", "Cash"}
			data.Clear()
			Dim rnd As New Random()
			For i As Integer = 0 To sNames.GetUpperBound(0)
				data.Rows.Add(New Object() {i + 1, sNames(i,0), sNames(i,1), sType(i Mod 3), rnd.Next(-1, 1), rnd.Next(10000) * 0.01})
			Next i
		End Sub

		Private Sub InitGridColumns()
			gridControl1.DataSource = data
			gridControl1.DefaultView.PopulateColumns()
			gridView1.Columns("Payment Amount").DisplayFormat.FormatType = DevExpress.Utils.FormatType.Numeric
			gridView1.Columns("Payment Amount").DisplayFormat.FormatString = "c"
			gridView1.BestFitColumns()
		End Sub

		Private Sub InitDataEvents()
			AddHandler data.RowDeleting, AddressOf data_RowDeleting
		End Sub

		Private Sub data_RowDeleting(ByVal sender As Object, ByVal e As DataRowChangeEventArgs)
			rowsSizes.Remove(e.Row)
		End Sub

		Private Sub gridView1_CalcRowHeight(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Grid.RowHeightEventArgs) Handles gridView1.CalcRowHeight
			Dim rowHeight As Object = rowsSizes(gridView1.GetDataRow(e.RowHandle))
			If rowHeight IsNot Nothing Then
				e.RowHeight = Convert.ToInt32(rowHeight)
			Else
				e.RowHeight = defaultRowHeight
			End If
		End Sub

		Private resizeStartPos As Integer
		Private resizingRowHandle As Integer = -1

		Private Sub gridView1_MouseDown(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles gridView1.MouseDown
			Dim hi As DevExpress.XtraGrid.Views.Grid.ViewInfo.GridHitInfo = gridView1.CalcHitInfo(e.X, e.Y)
			If hi.HitTest = DevExpress.XtraGrid.Views.Grid.ViewInfo.GridHitTest.RowEdge Then
				resizeStartPos = e.Y
				resizingRowHandle = hi.RowHandle
			End If
		End Sub

		Private Sub gridView1_MouseUp(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles gridView1.MouseUp
			If resizingRowHandle >= 0 Then
				resizeStartPos = e.Y - resizeStartPos
				Dim rowSize As Object = rowsSizes(gridView1.GetDataRow(resizingRowHandle))
				Dim newRowSize As Integer
				If rowSize IsNot Nothing Then
					newRowSize = Convert.ToInt32(rowSize)
				Else
					newRowSize = defaultRowHeight
				End If
				newRowSize += resizeStartPos
				If newRowSize < defaultRowHeight Then
					newRowSize = defaultRowHeight
				End If
				rowsSizes(gridView1.GetDataRow(resizingRowHandle)) = newRowSize
				resizingRowHandle = -1
			End If
		End Sub

	End Class
End Namespace
